libclamav/regex_list.c

```diff
@@ -630,21 +706,33 @@ static int add_newsuffix(struct regex_matcher *matcher, struct regex_list *info,
     new->pattern = MPOOL_MALLOC(matcher->mempool, sizeof(new->pattern[0]) * len);
     if (!new->pattern) {
         MPOOL_FREE(matcher->mempool, new);
+
         cli_errmsg("add_newsuffix: Unable to allocate memory for new->pattern\n");
-        return CL_EMEM;
+        ret = CL_EMEM;
+        goto done;
     }
-    for (i = 0; i < len; i++)
+    for (i = 0; i < len; i++) {
         new->pattern[i] = suffix[i]; /*new->pattern is short int* */
+    }
 
     new->customdata = info;
     new->virname    = NULL;
     if ((ret = cli_ac_addpatt(root, new))) {
         MPOOL_FREE(matcher->mempool, new->pattern);
         MPOOL_FREE(matcher->mempool, new);
-        return ret;
+        goto done;
     }
-    filter_add_static(&matcher->filter, (const unsigned char *)suffix, len, "regex");
-    return CL_SUCCESS;
+
+    if (filter_add_static(&matcher->filter, (const unsigned char *)suffix, len, "regex") < 0) {
+        MPOOL_FREE(matcher->mempool, new->pattern);
+        MPOOL_FREE(matcher->mempool, new);
+        cli_errmsg("add_newsuffix: Unable to add filter\n");
+        ret = CL_ERROR;
+        goto done;
+    }
+
+done:
+    return ret;
```

Comment:
Can I trouble you to do something like this after the `done:`:

```c
if (CL_SUCCESS != ret) {
    if (NULL != new) {
        if (NULL != new->pattern) {
            MPOOL_FREE(matcher->mempool, new->pattern);
        }
        MPOOL_FREE(matcher->mempool, new);
    }
}
```

Then remove the copy-paste free's from the error conditions for:
```c
    if (!new->pattern) {
```
```c
    if ((ret = cli_ac_addpatt(root, new))) {
```
```c
    if (filter_add_static(&matcher->filter, (const unsigned char *)suffix, len, "regex") < 0) {
```

Other than this one thing, this PR looks great to me.  Everything checked out in testing.

Reply:
good idea, done